### PR TITLE
E2E: handling extra files

### LINKED ISSUE #23987: bindata for examples, docs, and testing the docs
Kubectl.go, examples.go, and other e2e tests need the repository root.  Overall, We have ~ 30 different dependencies (`It("...")`) in the e2es referencing files from examples/ docs/ ...  This is specified via `repo-root` at the moment.

**Problem**
- This makes the shipped binary `e2e.test` unusable for `Conformance` testing,  since some Conformance specs rely on files that ship w/ kube.
-  This also forces hacky bootstrappy things like needing to make a tarball, clone a repo, check out a commit -  just to run the e2es.

**Goal**

These can all be eliminated permanantly, making the binary completely self sufficient for e2e testing.

**Solution**

Looks like its easy to serialize docs/examples as code: 

```
➜  kubernetes git:(kubectl-files) ✗  ~/Development/gopath/bin/go-bindata -o docs.go ./docs 
➜  kubernetes git:(kubectl-files) ✗  ~/Development/gopath/bin/go-bindata -o examples.go ./examples
➜  kubernetes git:(kubectl-files) ✗ ls -altrhy docs.go 

```

and you get some minified go files...

```
➜  kubernetes git:(kubectl-files) ✗ ls -altrh docs.go 
-rw-r--r--  1 jayunit100  staff    74K Apr  7 14:03 docs.go
➜  kubernetes git:(kubectl-files) ✗ ls -altrh examples.go 
-rw-r--r--  1 jayunit100  staff    52K Apr  7 14:03 examples.go
```

Then you can call functions on those files in the tests, telling them to create the assets, by calling functions in the generated files that recreate the static files:

```
107 
108 func examplesReadmeMd() (*asset, error) {
109         bytes, err := examplesReadmeMdBytes()
110         if err != nil {
111                 return nil, err
112         }
113 
114         info := bindataFileInfo{name: "examples/README.md", size: 2368, mode: os.FileMode(420), modTime: time.Unix(1458824586, 0)}
115         a := &asset{bytes: bytes, info: info}
116         return a, nil
117 }
```

Should this be part of the build, rather than needing `--repo-root` for the e2es ? That way they just bootstrap themselves using the `Asset(...)` thingy. 

cc @kubernetes/sig-testing  .....

### LINKED ISSUE #66649: refactor test/e2e/framework
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:

The E2E framework has grown organically over time. As a result, it is both hard to maintain and close to impossible to use outside of Kubernetes itself.

**What you expected to happen**:

The goal is to make it possible that other projects can:
- import test/e2e/framework as a vendor dependency without pulling in support for cloud providers and any tests (because of the huge dependency chains that a complete import causes)
- import support for cloud providers and tests separately if desired
- add custom tests without modifying the framework
- create and run a Ginkgo suite

One potential user for this is the Kubernetes CSI project, which currently relies on tests in Kubernetes to test the out-of-tree sidecar containers. Testing the "canary" sidecar containers is done manually by developers. There is an on-going effort to unify all sidecar containers in one repo (https://docs.google.com/document/d/1xC0Cm11WTebT9m_9yZbfzud24GJRxlp55gdR5QaMHmM/edit), which then could also import and run the E2E testing in TravisCI.

Other users are CSI driver developers.

For a full description of this proposal including an implementation plan, see this Google doc: https://docs.google.com/document/d/1lPG264ZlwPbIuFeP9P0-c0vXdmCC7wzwaapf53BbXBA/edit?usp=sharing

### PR DESCRIPTION
**Pull request by pohly**
**What this PR does / why we need it**:

There is an on-going effort to make test/e2e.test self-contained by building all files that are needed at runtime into the binary via bindata. Tests which support that do so by retrieving files from the test/e2e/generated package. The downside of that approach is that such tests only work in Kubernetes itself. As described in issue #66649, making tests reusable outside of Kubernetes is useful, for example for testing other CSI drivers.

This PR adds an API (test/e2e/framework/testfiles) that decouples tests from the actual code that retrieves files. In the test/e2e test suite, access via --repo-root is kept in addition to retrieval from bindata, so everything works as before.

The goal is *not* to get rid of --repo-root. However, by converting some tests to use the new API, those tests no longer depend on --repo-root, so we get a step closer to that goal - if it still is a goal. Some tests depend on extra files (for example, test/e2e/framework/ingress/ingress_utils.go for providing the optional secret.yaml) which cannot be built into the binary.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Partial fix #66649 (main goal)
Partial fix #23987 (side effect)

**Special notes for your reviewer**:

This depends on PR #68187 getting merged first. The first commit here is from that PR.

This is a subset of PR #68483.

**Release note**:

```release-note
NONE
```

/sig testing

### CODE REVIEW

File: test/e2e/framework/testfiles/testfiles.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+	DescribeFiles() string
+}
+
+// ReadOrDie tries to retrieve the desired file content from
+// one of the registered file sources. In contrast to FileSource, it
+// will either return a valid slice or abort the test, i.e. the caller
+// doesn't have to implement error checking.
+func ReadOrDie(filePath string) []byte {
+	if len(filesources) == 0 {
+		ginkgo.Fail(fmt.Sprintf("No file sources registered (yet?), cannot retrieve test file %s.", filePath))
```
[the remark this answers is not in the document]
Comment (timothysc) [reply test/e2e/framework/testfiles/testfiles.go]: After looking at the call sites I think you wrap everything into an error handler and pass in a closure to call the ginkgo with the error without you having to take on owning ginkgo details in this sub-package.

**Review by timothysc — CHANGES_REQUESTED** [review]
Minor comments + 2 commits, please. 

1 for changes 
1 for auto-generated.

File: test/e2e/framework/testfiles/testfiles.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/onsi/ginkgo"
+)
+
+var filesources = make([]FileSource, 0, 5)
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/e2e/framework/testfiles/testfiles.go]: No particular reason, I just thought that pre-allocating some memory might be useful. But I agree, that's something that the first append() can also do, so I'll make it nil.

File: test/e2e/framework/testfiles/testfiles.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pohly) [reply test/e2e/framework/testfiles/testfiles.go]: This is part of test/e2e/framework, which already is tied to Ginkgo, so it seemed okay to use that also here. It makes the API a bit simpler. What about just calling `panic`? The effect will be the same (also get's caught by Ginkgo and logged).

It's not just `ReadOrDie`, `Exists` also fails on errors to simplify the call sites.

**Comment by pohly** [comment]
@timothysc I force-pushed two commits, one for the source code changes and one for the Bazel BUILD files. I kept the test/e2e/examples.go changes because if we merge all of this at once, we do not need PR #68187 anymore.

This version uses `panic` in `test/e2e/framework/testfiles` instead of Ginkgo. I'd prefer to keep the API simple, but don't feel strongly about it and can also introduce an additional parameter which handles the errors.

Here's output for a test that triggers such a panic because the file name was changed such that the file isn't found. I've shortened the list of files. The original error lists all of the builtin files:
```
•! Panic in Spec Setup (BeforeEach) [6.161 seconds]
[sig-cli] Kubectl client
/nvme/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/kubectl/framework.go:22
  [k8s.io] Update Demo
  /nvme/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:694
    should do a rolling update of a replication controller  [Conformance] [BeforeEach]
    /nvme/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:699

    Test Panicked
    Test file "test/fixtures/doc-yaml/user-guide/update-demo/nautilus-xxx-rc.yaml.in" was not found.
    Test files are expected in "/nvme/gopath/src/k8s.io/kubernetes/cluster/..".
    The following files are built into the test executable via gobindata. For questions on maintaining gobindata, contact the sig-testing group.
       test/e2e/testing-manifests/flexvolume/dummy
       test/e2e/testing-manifests/flexvolume/dummy-attachable
....
       test/images/webhook/patch_test.go
       test/images/webhook/pods.go
       test/images/webhook/scheme.go
    
    /usr/lib/go-1.10/src/runtime/panic.go:502

    Full Stack Trace
    	/usr/lib/go-1.10/src/runtime/panic.go:502 +0x229
    k8s.io/kubernetes/test/e2e/framework/testfiles.ReadOrDie(0xc421f40410, 0x45, 0x2, 0xc421f40410, 0x45)
    	/nvme/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/testfiles/testfiles.go:90 +0x2e6
    k8s.io/kubernetes/test/e2e/kubectl.glob..func2.6.1()
    	/nvme/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/kubectl/kubectl.go:295 +0xad
    k8s.io/kubernetes/vendor/github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc420f4f800, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    	/nvme/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/e2e.go:306 +0x1fc
    k8s.io/kubernetes/test/e2e.TestE2E(0xc42056fa40)
    	/nvme/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/e2e_test.go:65 +0x2b
    testing.tRunner(0xc42056fa40, 0x45c7fe8)
    	/usr/lib/go-1.10/src/testing/testing.go:777 +0xd0
    created by testing.(*T).Run
    	/usr/lib/go-1.10/src/testing/testing.go:824 +0x2e0
```

File: test/e2e/framework/testfiles/testfiles.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (timothysc) [reply test/e2e/framework/testfiles/testfiles.go]: Passing via a well defined, simple closure it not too pesky and would meet the goal of minimizing the proliferation of ginkgo dependencies and being about to have the failure be clearly visible from the call sites.

**Comment by pohly** [comment]
@timothysc I pushed a version which takes a callback that gets invoked on failures. To make the API still easy to use, the signature is exactly that of ginkgo.Fail. Otherwise callers would have to use a closure.

Some callers can handle an error and don't use Ginkgo. For those I added a `testfiles.Read` which returns the error.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve 

Thanks for doing this!

**Comment by pohly** [comment]
@timothysc I had to rebase because https://github.com/kubernetes/kubernetes/commit/60f736ed3d617f4770dc76657dd75a31bffb03f0 introduced a new direct call to the `generated` package. Please add lgtm again.

There's also some test failure in pull-kubernetes-integration which I don't understand ("test-cmd run_rs_tests" fails) - a flake or does this PR really cause that test to fail now?

**Pull request merged**